packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1287,10 +1269,45 @@ async function renderToHTMLOrFlightImpl(
     return new RenderResult(await streamToString(response.stream), options)
   } else {
     // We're rendering dynamically
+
+    if (
+      process.env.NODE_ENV === 'development' &&
+      renderOpts.setAppIsrStatus &&
+      // The type check here ensures that `req` is correctly typed, and the
+      // environment variable check provides dead code elimination.
+      process.env.NEXT_RUNTIME !== 'edge' &&
+      isNodeNextRequest(req)
+    ) {
+      const setAppIsrStatus = renderOpts.setAppIsrStatus
+      req.originalRequest.on('end', () => {
+        if (!requestStore.usedDynamic && !workStore.forceDynamic) {
```

Comment:
lgtm, but this'll be clearer if we move it after we instantiate the request store